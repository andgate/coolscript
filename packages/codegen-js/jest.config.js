const base = require('../../jest.config.base.js')
const { name: packageName } = require('./package.json')

module.exports = {
  ...base,
  name: packageName,
  displayName: packageName,
  rootDir: './',
  moduleNameMapper: {
    '^@coolscript\/codegen-js$': `<rootDir>/src`
  }
}
