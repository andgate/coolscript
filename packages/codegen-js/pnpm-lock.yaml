lockfileVersion: 5.3

specifiers:
  '@coolscript/syntax-concrete': ^0.1.0

devDependencies:
  '@coolscript/syntax-concrete': 0.1.0

packages:

  /@coolscript/syntax-concrete/0.1.0:
    resolution: {integrity: sha512-fSNmdIUNRbRWsDYjXM+W3475RU3Kt0A97u84EwHG8u9/oKZgl4zwSSzHSwm0K/poexulULS8NFSmA8IPoXit8g==}
    dev: true
