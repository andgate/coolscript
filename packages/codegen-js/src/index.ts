export * from './coolscript/codegen/js'
