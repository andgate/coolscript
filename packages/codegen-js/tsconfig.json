{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "noEmit": true
  },
  "include": ["src", "__tests__"],
  "exclude": ["lib", "node_modules"]
}
