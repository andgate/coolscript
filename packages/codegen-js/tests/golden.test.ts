import { examples } from '@coolscript/examples'
import { codegenJS } from '@coolscript/codegen-js'

const exampleKeys: string[] = Object.keys(examples)

describe('CoolScript JS Codegen Golden Test Suite', () => {
  test.each(exampleKeys)('Generating example %p', (exampleKey) => {
    const example = examples[exampleKey]
    const parseResult = codegenJS(example)
    expect(parseResult).toMatchSnapshot(exampleKey)
  })
})
