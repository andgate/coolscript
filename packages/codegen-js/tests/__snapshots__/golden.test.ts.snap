// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoolScript JS Codegen Golden Test Suite Generating example "example1": example1 1`] = `
Object {
  "source": "(function () {
  const id = function (x) {
    return x;
  };
  const constant = function (a, b) {
    return a;
  };
  const createFoo = function (x) {
    return {
      name: [\\"F\\", \\"o\\", \\"o\\", \\"b\\", \\"a\\", \\"r\\"],
      value: x
    };
  };
  const bar = 999;
  return (function () {
    let baz = id(bar);
    let clientName = createFoo(baz).name;
    constant(baz, bar);
    return clientName;
  })();
})();
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "example2": example2 1`] = `
Object {
  "errors": Array [
    [Error: Parser failed! Error message: Syntax error at line 12 col 18:

      j = s.length -
                   ^
Unexpected keyword token: "-". Instead, I was expecting to see one of the following:

A ";" based on:
    statement_list_tail → _ ● ";" _ statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A newline token based on:
    _ → _ ● %newline _
    index_access_term → bterm ● _ "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A comment token based on:
    _ → _ ● %comment _
    index_access_term → bterm ● _ "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "(" based on:
    call_term → bterm _ ● "(" _ call_arguments _ call_term$ebnf$1 ")"
    bterm →  ● call_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "." based on:
    member_access_term → bterm _ ● "." _ identifier_token
    bterm →  ● member_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "[" based on:
    index_access_term → bterm _ ● "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "}" based on:
    block_statement → "{" _ block_statement$ebnf$1 _ block_statement$ebnf$2 ● "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A ";" based on:
    block_statement$ebnf$2$subexpression$1 →  ● ";" _
    block_statement$ebnf$2 → block_statement$ebnf$2 ● block_statement$ebnf$2$subexpression$1
    block_statement → "{" _ block_statement$ebnf$1 _ ● block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
],
  ],
  "source": null,
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "let1": let1 1`] = `
Object {
  "source": "(function () {
  const x = 1;
  return x;
})();
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "let2": let2 1`] = `
Object {
  "source": "(function () {
  const x = 1;
  return x;
})();
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "let3": let3 1`] = `
Object {
  "source": "(function () {
  const x = 1;
  const y = 2;
  return [x, y];
})();
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "number1": number1 1`] = `
Object {
  "source": "0;
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "number2": number2 1`] = `
Object {
  "source": "1234567890;
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "number3": number3 1`] = `
Object {
  "source": "1234567890.0123458;
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string1": string1 1`] = `
Object {
  "source": "\\"\\";
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string2": string2 1`] = `
Object {
  "source": "\\"\\";
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string3": string3 1`] = `
Object {
  "source": "\\"Hello world!\\";
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string4": string4 1`] = `
Object {
  "source": "\\"Hello World!\\";
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string5": string5 1`] = `
Object {
  "source": "\\"The quick brown fox jumps 'over' the lazy dog.\\";
",
}
`;

exports[`CoolScript JS Codegen Golden Test Suite Generating example "string6": string6 1`] = `
Object {
  "source": "\\"The quick brown fox \\\\\\"jumps\\\\\\" over the lazy dog.\\";
",
}
`;
